tests/chocolatey-tests/Hooks.Tests.ps1

```diff
@@ -0,0 +1,93 @@
+Import-Module helpers/common-helpers
+
+Describe "hooks tests" -Tag Chocolatey, Hooks {
+    BeforeDiscovery {
+        $Flags = @(
+            @{ Flag = '' ; RunsHooks = $true ; Command = 'install' }
+            @{ Flag = '--skip-powershell' ; RunsHooks = $false ; Command = 'install' }
+            @{ Flag = '--skip-hooks' ; RunsHooks = $false ; Command = 'install' }
+
+            @{ Flag = '' ; RunsHooks = $true ; Command = 'uninstall' }
+            @{ Flag = '--skip-powershell' ; RunsHooks = $false ; Command = 'uninstall' }
+            @{ Flag = '--skip-hooks' ; RunsHooks = $false ; Command = 'uninstall' }
+
+            @{ Flag = '' ; RunsHooks = $true ; Command = 'upgrade' }
+            @{ Flag = '--skip-powershell' ; RunsHooks = $false ; Command = 'upgrade' }
+            @{ Flag = '--skip-hooks' ; RunsHooks = $false ; Command = 'upgrade' }
+
+        )
+    }
+    BeforeAll {
+        Initialize-ChocolateyTestInstall
+        # Add the hooks package to the test install so that it is available in each snapshot.
+        $null = Invoke-Choco install scriptpackage.hook --version 1.0.0
+        New-ChocolateyInstallSnapshot
+    }
+
+    AfterAll {
+        Remove-ChocolateyTestInstall
+    }
+
+    Context '<Command> hooks with flag: <Flag>' -ForEach $Flags {
+        BeforeAll {
+            Restore-ChocolateyInstallSnapshot
+            $Package = 'upgradepackage'
+
+            if ($Command -ne 'install') {
+                $ver = if ($Command -eq 'upgrade') {
+                    '1.0.0'
+                }
+                else {
+                    '1.1.0'
+                }
+                $null = Invoke-Choco install $Package --version $ver --no-progress
+            }
+
+            $Output = Invoke-Choco $Command $Package $Flag --no-progress
+        }
+
+        # Uninstall/Upgrade exit -1: https://github.com/chocolatey/choco/issues/2822
+        It "Exits with Success (0,-1)" {
+            $Output.ExitCode | Should -BeIn @(0, -1) -Because $Output.String

```

Comment:
This will have false positives.

What would happen if the install command fails with exit code `-1`?

Would be better to separate this test into two separate tests where one is skipped on install, and the other only runs on install instead of allowing both `0` and `-1` values for all commands.

Reply:
Good points. I've extracted the expected value into the ForEach and so we can be explicit about which one we expect.